core/selectionoptimizer.js

```diff
@@ -0,0 +1,57 @@
+( function() {
+	CKEDITOR.dom.selection.prototype.optimizeInElementEnds = function() {
+		if ( this.isFake ) {
+			return;
+		}
+
+		var range = this.getRanges()[ 0 ];
+
+		if ( range.isCollapsed ) {
+			return;
+		}
+
+		if ( range.startContainer.equals( range.endContainer ) ) {
+			return;
+		}
+
+		if ( !shouldOptimize( range ) ) {
+			return;
+		}
+
+		range.shrink( CKEDITOR.SHRINK_TEXT );
+
+		preventRecurrency( this.root.editor, range );
+
+		range.select();
+	};
+
+	// Returns whether any condition is met:
+	// - range starts at the end of an element.
+	// - range ends at the beginning of an element.
+	// - one end of range is in text, and another is not.
+	function shouldOptimize( range ) {
+		if ( range.endOffset === 0 ) {
+			return true;
+		}
+
+		var startsInText = range.startContainer.type === CKEDITOR.NODE_TEXT,
+			limit = startsInText ? range.startContainer.getLength() : range.startContainer.getChildCount();
+
+		if ( range.startOffset === limit ) {
+			return true;
+		}
+
+		return startsInText ^ range.endContainer.type === CKEDITOR.NODE_TEXT;
```

Comment:
I'm not a big fan of using bitwise operators, as they are not commonly used in JS usually when seeing them have to remember myself what they do. But alternative conditional isn't very readable, maybe instead we can just add a comment `// XOR operator.` WDUT?

Reply:
It's probably more individual and depends who reads the code. For me alternative conditional is more readable and easier to parse, but I understand it isn't for others.

When it comes to comment, something like `// XOR operator.` is a little too laconic - most people looking on the code knows it's XOR but still will have to check how it works exactly. So in case of a comment you will need more explicit one which describes the condition, e.g.:

```
// Check if range starts or ends in text (not both) using XOR operator.
```

So you end up with two lines, one with actual code and the other one describing it. And on the other end you have alternative conditional which is one line and kind of self explanatory. I don't have very strong opinion here so decide which on you prefer.